### PR TITLE
win_powershell - Added module for executing powershell scripts

### PR DESCRIPTION
**Pull request by jborean93**
##### SUMMARY
Add a module that can be used to execute PowerShell script like `win_shell`. Where this differs with `win_shell` is it adds the outputted objects from the powershell script as a structured return value like:

```powershell
- ansible.windows.win_powershell:
    script: |
      Get-Service | Select Name, @{N='Status'; E={ $_.Status.ToString() }}
  register: service_info

- ansible.windows.win_service:
    name: '{{ item.Name }}'
    state: started
  with_items: '{{ service_info.output }}'
  when: '{{ item.Status != "Running" }}'
```

It also adds a few more benefits like

* Each individual stream record in the registered output
    * Error records can be parsed as a structured object
    * Warning/Verbose/Debug records are a list of string
    * Information records are also a structured object
* Can easily pass in objects as parameters rather than trying to embed them as a string that `win_shell` demands
    * Module options are converted to json then back from json as .NET objects so you can easily pass in a list or dict as a parameter
* No need for starting a new powershell process
    * No more `-EncodedCommand` that ticks up some AVs
* Not really a benefit over `win_shell` but you can use this for both Windows PowerShell and PowerShell using `executable`
* Can have the script modify some of the returned module output through the `$Ansible` var
    * `$Ansible.Changed` sets the changed status of the module
    * `$Ansible.Failed` sets the failed status of the module
    * `$Ansible.Result` can set more structured returned values in whatever format the user desired
    * `$Ansible.Tmpdir` can be used by the script to get the Ansible configured tempdir

Ultimately this is similar to `win_dsc` where you can write your own mini modules in a task instead of manually converting the output to json, registering the stdout, converting the stdout back from json and so on.

There are some downsides to this module:

* We serialize the whole output, there could be cases when it brings `ConvertTo-Json` to its knees and there is nothing we can do about that
    * We are bound to get some issues raised for this but I'm not sure how much we can do except stress the importance to check  what your scripts output
    * Maybe we need to add a final passthru sanity check on the returned objects for known offenders, _cough System.Type cough_.
* Objects are pure objects and not formatted to the host output
    * People will probably think that an `PSCustomObject` should look like the table or list format but in reality we will output the object and their properties like a dict
    * Once again this is really just a behaviour thing, if they want this behaviour then they need to use `win_shell`
The downside with this module will inevitably stem from people expecting it to behave like the `powershell.exe` console
* `Write-Host` won't write to the `output` return value
    * It is captured in `host_out` but some people may expect it to be `normal` output
    * We could try and mix them together but personally I think that's a mistake and we should keep it separate

####### Error and Information Record structure

We currently convert these records into a more structured format to avoid issues with json serializing complex types. It does drop some extra type specific information but IMO it's already providing enough information as is. Maybe we can simplify it even more by dropping some of the return values. Should we keep the behaviour as it is in the PR or change it in some way?

####### $Ansible variable

We inject a `PSCustomObject` as the `$Ansible` variable that the script can interact with. This variable can be used by the script to add explicitly named returned values or set the changed/failed state based on their own logic. Should we even bother to offer this? It's not complicated but it might just be relegated to one of those features no one knows about and thus isn't used at all.

####### Default $ErrorActionPreference

The current `$ErrorActionPreference` is the default in PowerShell (`Continue`) which means any error records are added to the error stream but the script continues on. We have an open issue #122 that has asked to change the default to `Stop` as that is safer when it comes to reporting failures. Should we explicitly set this to `Stop` and expose a module option (or just have them manually set it in their script) to change it to something else?

I personally think we should keep it on `Continue` as anyone making PowerShell functions should be aware of this behaviour. It's really only the people new to PowerShell that might be aware of it. Currently the `failed` state of the module is set to `True` if the pipeline sets the `HadErrors` property. AFAIK this is set to `True` if there are any error records which would alleviate some of the concerns from #122. I need to play with that a bit more to see when `HadErrors` is set to `True` to give some more detail here.

####### Default location/chdir

The current logic is to use whatever the module location (or new process) is set at. This can lead to different defaults depending on the connection plugin used or even if you are becoming an elevated or limited account. I personally think we should always set the default location to `C:\Users\<username>` (`$env:USERPROFILE`) but wondering what others thing.

##### TODO

- [x] Add lots of tests, especially around error handling and host output
- [x] Fix sanity issues with docs
- [x] Add more doc examples
- [x] Add more info the docs in regards to this vs `win_shell`
- [x] Figure out how to set the location, should we default to a certain dir as the various connections differ in what their defaults are

Fixes https://github.com/ansible-collections/ansible.windows/issues/122

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
win_powershell

### CODE REVIEW

**Review by briantist — COMMENTED** [review]
This is awesome, so happy to see this!

Going to comment on the some of the questions and info raised in your original post too.

---

> Error and Information Record structure
> Maybe we can simplify it even more by dropping some of the return values. Should we keep the behaviour as it is in the PR or change it in some way?

Looks good so far (but I haven't executed it yet). Saw some stuff in the information stream that could probably be dropped (thread IDs, feels like that probably can't be used after returned unless it's good for correlation with logs?).

This is a little out there maybe, but one thing that could be cool later on is to pair this with `test_plugins`. If we think it would be useful to look for certain types of failures or exceptions, and the results have the right fields, that'll be possible already, but test plugins can make that really nice.

As an example, I did this (internally at work) for the AWS modules in Ansible. I was trying to do retries when they hit throttling errors. So I'd use some jinja like this:
```yaml
register: task_result
until: not (task_result.error is defined and task_result.error.code is defined and task_result.error.code == 'Throttled')
```
I made a simple test plugin called `aws_error` that could test for the codes, so I could do this instead:

```yaml
until: task_result is not aws_error('Throttled')
```

And then in the same file I defined a few specific tests that end up being one liners:
```python
def aws_api_throttled(result):
    '''tests to see if a task result is an AWS call that failed due to throttling'''
    return aws_error(result, error_code='Throttling')
```
So I can also do:
```yaml
until: task_result is not aws_api_throttled
```

Anyway point is, as long as the fields are there, stuff like that is possible which is very nice, so I appreciate the structured nature and not having to regex or string match error messages :)

---

> $Ansible variable
> Should we even bother to offer this? It's not complicated but it might just be relegated to one of those features no one knows about and thus isn't used at all.

Yes please offer this! This is one of the defining features of this module imo. 

---

> Default $ErrorActionPreference

As I mentioned in #122 I agree that keeping it to PowerShell's default makes the most sense; we should consider that one of the utilities of using a module like this is not just for direct code sent in but execution of existing code via `Import-Module` (which isn't always as useful with `win_shell` without wrappers to get info in and out), and unfortunately that code may fail unexpectedly with a different default.

---

> Default location/chdir
>
> The current logic is to use whatever the module location (or new process) is set at. This can lead to different defaults depending on the connection plugin used or even if you are becoming an elevated or limited account. I personally think we should always set the default location to C:\Users\<username> ($env:USERPROFILE) but wondering what others thing.

I think letting the directory be whatever it is is probably fine for the default case, and adding a `chdir` option to allow for setting the working directory is sufficient for when it matters (and gives more flexibility).

Was trying to think if `$env:USERPROFILE` would cause any issues in particular situations (roaming profiles, redirected folders, system/network accounts), but I can't immediately think of any problem with it.

---

Re: `Write-Host` behavior: totally agree with your current choice: it shouldn't be mixed up with the output stream.

Re: "Objects are pure objects and not formatted to the host output": same, I wouldn't at all expect the host formatted output, but even for people who might, it's quite easy to point them at `win_shell`

---

> We serialize the whole output, there could be cases when it brings ConvertTo-Json to its knees and there is nothing we can do about that
> We are bound to get some issues raised for this but I'm not sure how much we can do except stress the importance to check what your scripts output
> Maybe we need to add a final passthru sanity check on the returned objects for known offenders, _cough_ System.Type _cough_.

I really felt this when writing [`win_psmodule_info`](https://github.com/ansible-collections/community.windows/blob/main/plugins/modules/win_psmodule_info.ps1). There's stuff you expect will be tricky like circular references, and then there's things you wouldn't expect to have to treat specially like `DateTime` and `System.Version` to get sane output.

This is going off on a tangent, but I wonder if trying to offer "saner" serialization for Ansible is a good thing. The way I'd see this working is catching common use cases (`[Type]`, `[Version]`, `[DateTime]`, etc.), and having it be the default, but maybe giving a module option for "raw" serialization for those who want to handle it themselves and don't like the choices made in the "magic" serialization.

If this functionality were in a module util, it could be used by other module authors (_ahem_ me) to not have to reimplement (as much of) that.

Admittedly, it's a quagmire, and will never catch everything, but.. a guy can dream.

---

### Suggestion

So I only really have one behavior change request and that's around check mode. I definitely agree, that like `win_command` and `win_shell`, we can't blindly execute in check mode because we don't know what could be changed, but with the `$Ansible` variable and possibly by using `-WhatIf`/`$WhatIfPreference` we have some opportunity for check mode support.

I'm kind of a stickler for Ansible plays being written in a way that works properly in check mode, so I try to avoid modules like this or `win_shell` but when I have to I usually do something like this:
```yaml
        win_shell: |
          $check_mode = [bool]::Parse('{{ ansible_check_mode }}')

          if (-not ($SomeCondition) {
              if (-not $check_mode) {
                Do-Thing
              }
              '__changed__'
          } else {
              '__ok__'
          }
        register: status
        check_mode: no
        failed_when: status.stdout is not defined
        changed_when: status.stdout is search('__changed__')
```

The `$Ansible` variable already solves the status difficulties. Although it may not get much use, I would personally prefer two modes of being able to execute in check mode (**both opt-in**):
- One that only populates `$Ansible.CheckMode`
- One that also set `$WhatIfPreference`

Actually maybe that last one could be handled by somewhat with `[CmdletBinding(SupportsShouldProcess)]` at the top of your script, and passing `parameters['WhatIf'] = True` 🤔

The first mode could also be worked around, just wondering if it makes sense to provide some kind of first class support for it.

---

Ok anyway, super excited for this, love that all the input streams are separated and this allows for a real nice intermediate step between `win_shell` (annoying to write around check mode and status) and writing a full module (this can be hard to keep maintained and supported in an organization).

File: plugins/modules/win_powershell.ps1 [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+        pipeline_iteration_info = $_.PipelineIterationInfo
+    }
+})
+
+'debug', 'verbose', 'warning' | ForEach-Object -Process {
+    $module.Result.$_ = @($ps.Streams.$_ | Select-Object -ExpandProperty Message)
+}
+
+# TODO: Will this fail on pre v5 as they don't have the Information stream.
+$module.Result.information = @($ps.Streams.Information | ForEach-Object -Process {
```
[the remark this answers is not in the document]
Comment (jborean93) [reply plugins/modules/win_powershell.ps1:369]: Yea I think what's there will be fine as we aren't in strict mode and invalid properties just turn into `$null` therefore achieving the same thing. Just a note to keep in mind when I go further into testing.

File: plugins/modules/win_powershell.ps1 [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+        # When testing a path like Cert:\LocalMachine\My, System.IO.File will
+        # not work, we just revert back to using Test-Path for this
+        return Test-Path -Path $Path
+    }
+
+    if ([Int32]$attributes -eq -1) {
+        return $false
+    } else {
+        return $true
+    }
```
[the remark this answers is not in the document]
Comment (jborean93) [reply plugins/modules/win_powershell.ps1:226]: This is essentially a copy of what's in `Ansible.ModuleUtils.FileUtil.psm1` in Ansible but I didn't really want to continue adding deps on builtin utils for very small things. Essentially there are a few system files, like `c:\pagefile.sys` where Win32 may not be able to open a handle to the file. This breaks the normal checks like `Test-Path` so instead we use `GetAttributes` which in .NET is able to handle this with a certain workaround. Potentially the attribute check was a leftover when the code originally did `[FileInfo]'C:\missing\path'` which doesn't raise an exception but sets it to `-1`. It seems like I could simplify this so thanks for bringing it up.

This issue was initially reported in https://github.com/ansible/ansible/issues/30258.

**Comment by jborean93** [comment]
> Anyway point is, as long as the fields are there, stuff like that is possible which is very nice, so I appreciate the structured nature and not having to regex or string match error messages :)

Yea my logic was it could be nice to have this for something in the future and by setting a strict structure it's a lot easier to parse without having knowledge of the .NET structure. There's also the json serialization problem I've mentioned before. My main concern is if we are adding too much info, there's probably little reason to have `pipeline_iteration_info` or some of the exception fields but I just kept them there for now.

> Yes please offer this! This is one of the defining features of this module imo.

I definitely like the idea, my main concern is people just won't know about it so it's just extra work with little gain. It also makes things harder if we want to change how this works or just remove it in the future. The main reason I had it is as you have mentioned it gives a nice basic mini module in a task for some of the small pieces of work where a full module would be overkill. I'll keep it in for now but happy to see if there's any other improvements or changes you would be interested in.

> As I mentioned in #122 I agree that keeping it to PowerShell's default makes the most sense

Yea I'm definitely leaning towards not adjusting things and keeping it as the defaults, the PowerShell team obviously chose this for a reason :).

> I think letting the directory be whatever it is is probably fine for the default case, and adding a chdir option to allow for setting the working directory is sufficient for when it matters (and gives more flexibility).

The `location` parameter is what I had for this I just haven't plugged it in yet. I probably should just change it to `chdir` to make `(win_)command/shell` to keep things simple. My main qualm about bringing it up is that the `winrm` connection plugin uses `C:\Windows\<username>` and `psrp` uses `C:\Windows\<username>\Documents` as the default location making scripts that are dependent on relative paths break when swaping plugins. IMO if you are using relative paths in your code then you should be using `chdir` but it's nice to keep things consistent. I'll have to think about this some more and see if there's any more compelling reasons to choose one or the other.

> This is going off on a tangent, but I wonder if trying to offer "saner" serialization for Ansible is a good thing.

The issue here is the serialization work is done in [Ansible.Basic](https://github.com/ansible/ansible/blob/f533d46572113655a0a698beab4b38671744a458/lib/ansible/module_utils/csharp/Ansible.Basic.cs#L402-L423) so any changes there will take time to propagate through and now we have a a module dependent on Ansible versions. We could reduce the depth but that will break any modules that rely on returning deeply nested objects (with this being one of them). Even so I've found that 4-5 is the magic spot when it comes to certain types with a `System.Type` usually hanging the console at that number making a lower default quite limiting.

I'm not we will ever really adequately solve this problem and normally I would say it's just something the module authors should be aware of and deal with themselves. This module definitely makes it harder to say that as end users will be writing code that can output the data directly, or even without their knowing, causing a task to hang. I'll also have a further think about this one and see if we want to have a manual check for known offenders like `[Type]`, `[DateTime]` and so on.

> So I only really have one behavior change request and that's around check mode

Checking for the precence of `[CmdletBinding(SupportsShouldProcess=$true)]` is definitely a nice option that allows us to extend the check mode validation into the script itself. Shouldn't be too hard to parse the script as AST and check if that's present to control that logic. The only apprehension I would have with this is some of the gnarlier aspects of `-WhatIf` and how it is sent to any nested functions/cmdlets. I don't fully understand the rules but IIRC the `-WhatIf` state is dependent on whether the function is in the same module or not. This could be alleviated by also setting `$Ansible.CheckMode` so the script can either check `$Ansible.CheckMode` or `$PSCmdlet.ShouldProcess`.

I will have to check how this impacts the impact level advertised. Typically PowerShell prompts for confirmation if the impact is higher than the preference var which should fail in this module but it's something to check. Maybe we just need to also pass in `-Confirm:$false` if it's `SupportsShouldProcess`.

> One that only populates $Ansible.CheckMode

I'm not 100% sure how this one will work. It's the module itself that defines `$Ansible` and not the script. If we were to pass in the check mode information rather than just skip then we need some way for the script itself to advertise that it supports this.

Thanks for the review, there's definitely a few things I need to think about some more here.

File: plugins/modules/win_powershell.ps1 (hunk 2 above)
[the remark this answers is not in the document]
Comment (briantist) [reply plugins/modules/win_powershell.ps1:226]: That's good to know, I never ran into that before. Thanks for the context!

**Comment by briantist** [comment]
> The issue here is the serialization work is done in Ansible.Basic so any changes there will take time to propagate through and now we have a a module dependent on Ansible versions. 

Oh right I definitely wasn't proposing changing what's in `Ansible.Basic`, was thinking more along the lines of an alternate serialization that could be chosen, or maybe serialization is the wrong word, it's more like serialization prep, like I was doing in `win_psmodule_info`; basically changing some values around so they end up serializing properly. So it'd be more something module authors could opt into and choose to do before sending their output.

And for this module using that by default, with maybe an option the caller could choose to turn that off, and handle it all themself if they really wanted.

> Checking for the precence of [CmdletBinding(SupportsShouldProcess=$true)] is definitely a nice option that allows us to extend the check mode validation into the script itself. Shouldn't be too hard to parse the script as AST and check if that's present to control that logic.

Ah right, that should be easy to test for with the AST!

> The only apprehension I would have with this is some of the gnarlier aspects of -WhatIf and how it is sent to any nested functions/cmdlets.

But to your point of shouldprocess being... weird (and often poorly implemented), that's why I figured setting that would have to be opt-in and not automatic with check mode (unfortunately).

> I will have to check how this impacts the impact level advertised. 

I don't think `SupportsShouldProcess` alone changed the default impact level, you can set that separately within cmdletbinding with `ConfirmImpact=whatever`. But you're right it will be good to check because if the threshold is tipped it'll try to prompt.

---

Happy to help test some of this stuff too! 🙂

**Comment by nitzmahone** [comment]
I still haven't had time for a deep pass over this, but a few thoughts:
* **$Ansible** I also like this idea- seems like we could potentially hang all sorts of useful properties and methods off that in the future beyond just affecting the module result. Would it make more sense as an injected type accelerator/static accessor thing instead of a reserved varname? eg `[AnsibleModule]::Changed = $false` - could see this one going either way, just a thought...
* **ErrorActionPreference** I suspect non-PS people would prefer default `Stop` and PS people would prefer the matching default. Maybe expose it as a module arg? That'd at least let people use `module_defaults` if they really want to change the default behavior.
* **Default dir** This one's tough, as even Windows does `C:\Windows\system32` sometimes and `~` other times... If we want a sane default `~` seems reasonable, though makes me a little more nervous for `become` scenarios.

+1 to exposing `CheckMode` state on `$Ansible` (or whatever) as well, but yeah, leave the "what to do with it" up to the scripts- no sense in providing a false sense of security that all commands can be run with impunity in Check Mode when that's not always the case... :)

I'd also probably default to a fairly low default depth (eg 5-10) on `ConvertTo-Json` and have a knob to increase it.

**Comment by jborean93** [comment]
Thanks for the comments everyone.

> Would it make more sense as an injected type accelerator/static accessor thing instead of a reserved varname? 

Unfortunately for this to happen we would have to create a C# class through `Add-Type`. While we already create a custom C# type for the custom PSHost we would need to do this separately in the actual PS runspace we running the code in. This adds more time to the execution. Other problems with this approach

* Need to inject the static state values like check mode, into the C# definition
* Need to retrieve the results value after our script has run by running another script rather than just through the session state proxy
* The type is persisted for the entire process, not an issue now but could be once we start doing persistance

I think the simplest option is to keep the var as it's quite flexible and simple to access.

**Comment by nitzmahone** [comment]
Ah, fair points all... The static data thing wouldn't really be a problem with persistence, we'd just have to roll a TLS execution context for it (same way the object proxy evaluation context magic works), but yeah, just injecting it as a variable gets you that for free, and PS var scoping being what it is, that's probably good enough for most things.

**Review by nitzmahone — APPROVED** [review]
holy-has-to-reimplement-the-world Batman

**Comment by jborean93** [comment]
Thanks for the review @nitzmahone
